xgcm/transform.py

```diff
@@ -0,0 +1,225 @@
+"""
+Classes and functions for 1D coordinate transformation.
+"""
+import functools
+import numpy as np
+import xarray as xr
+from numba import jit, guvectorize, float32, float64, boolean
+
+
+@guvectorize(
+    [
+        (float64[:], float64[:], float64[:], boolean, float64[:]),
+        (float32[:], float32[:], float32[:], boolean, float32[:]),
+    ],
+    "(n),(n),(m),()->(m)",
+    nopython=True,
+)
+def _interp_1d_linear(phi, theta, target_theta_levels, mask_edges, output):
+    output[:] = np.interp(target_theta_levels, theta, phi)
+
+    if mask_edges:
+        theta_max = np.nanmax(theta)
+        theta_min = np.nanmin(theta)
+        for i in range(len(target_theta_levels)):
+            theta_lev = target_theta_levels[i]
+            if (theta_lev < theta_min) or (theta_lev > theta_max):
+                output[i] = np.nan
+
+
+def interp_1d_linear(phi, theta, target_theta_levels, mask_edges=False):
+    """
+    Vectorized interpolation of scalar phi to isosurfaces of scalar theta
+    along the final axis.
+
+    Parameters
+    ----------
+    phi : array_like
+        Array of shape (..., n), scalar field to be interpolated
+    theta : array_like
+        Array of shape (..., n), scalar field which defines the isosurfaces
+    target_theta_levels : array_like
+        Array of shape (m) specificying target isosurface levels
+    mask_edges : bool, optional
+        Determines how to handle theta values that exceed the bounds of
+        target_theta_levels. If False, fill with nearest valid values. If
+        True, fill with NaNs.
+
+    Returns
+    -------
+    phi_interp : array
+        Array of shape (..., m) of phi interpolated to theta isosurfaces.
+    """
+    return _interp_1d_linear(phi, theta, target_theta_levels, mask_edges)
+
+
+@guvectorize(
+    [
+        (float64[:], float64[:], float64[:], float64[:], float64[:], float64[:]),
+        (float32[:], float32[:], float32[:], float32[:], float32[:], float32[:]),
+    ],
+    "(n),(n),(n),(m),(m)->(m)",
+    nopython=True,
+)
+def _interp_1d_conservative(phi, theta_1, theta_2, theta_hat_1, theta_hat_2, output):
+    output[:] = 0
+
+    n = len(theta_1)
+    m = len(theta_hat_1)
+
+    for i in range(n):
+        # handle non-monotonic stratification
+        if theta_1[i] < theta_2[i]:
+            theta_min = theta_1[i]
+            theta_max = theta_2[i]
+        else:
+            theta_min = theta_2[i]
+            theta_max = theta_1[i]
+
+        for j in range(m):
+            if (theta_hat_1[j] > theta_max) or (theta_hat_2[j] < theta_min):
+                # there is no overlap between the cell and the bin
+                pass
```

Comment:
```python
elif theta_max == theta_min:
    output[j] = phi[i]
```

I think that should be enough, no?

Reply:
Yeah actually you are right I think. Let me try that.

Reply:
I just edited it to include `+=`.

Reply:
Yup that seems to work. Just pushed the changes.